Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,446 @@
+        }
+        
+        /// tries to find an available connection on the required `eventLoop`
+        private func findAvailableConnection(onRequired eventLoop: EventLoop) -> Int? {
+            self.connections.firstIndex(where: { $0.eventLoop === eventLoop })
+        }
+        
+        /// - precondition: `index` needs to be valid
+        mutating func leaseStreams(at index: Int, count: Int) -> Connection {
+            self.connections[index].lease(count)
+        }
+        
+        mutating func releaseStream(_ connectionID: Connection.ID) -> (Int, AvailableConnectionContext) {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to release a connection we do not know anything about")
+            }
+            let availableStreams = self.connections[index].release()
+            let context = AvailableConnectionContext(availableStreams: availableStreams, eventLoop: self.connections[index].eventLoop)
+            return (index, context)
+        }
+        
+        // MARK: Connection close/removal
+        
+        mutating func closeConnection(at index: Int) -> Connection {
+            self.connections[index].close()
+        }
+        
+        mutating func removeConnection(at index: Int) {
+            precondition(self.connections[index].isClosed, "We tried to remove a connection which is not closed")
+            self.connections.remove(at: index)
+        }
+
+        mutating func closeConnectionIfIdle(_ connectionID: Connection.ID) -> Connection? {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                // because of a race this connection (connection close runs against trigger of timeout)
+                // was already removed from the state machine.
+                return nil
+            }
+            guard self.connections[index].isIdle else {
+                // connection is not idle anymore, we may have just leased it for a request
+                return nil
+            }
+            return self.closeConnection(at: index)
+        }
+        
+        mutating func createNewConnectionByReplacingClosedConnection(at index: Int) -> (Connection.ID, EventLoop) {
+            precondition(self.connections[index].isClosed)
+            let newConnection = HTTP2ConnectionState(
+                connectionID: self.generator.next(),
+                eventLoop: self.connections[index].eventLoop
+            )
+
+            self.connections[index] = newConnection
+            return (newConnection.connectionID, newConnection.eventLoop)
+        }
+        
+        mutating func failConnection(_ connectionID: Connection.ID) -> (Int, FailedConnectionContext)? {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                return nil
+            }
```

Comment:
I think it would be useful, if we had a comment here, explaining why we might get this callback, and we know nothing about a connection with this id.